File: Codes/Order/blind.py
import numpy as np
from scipy.signal import convolve2d

def I_approximated(I,B):
    omega = [0.4,0.35,0.25]
    H4 = np.ones((3,3))/9

    rgb = []

    for k in range(3):
        Ik = I[:,:,k]*1
        Bk = B[:,:,k]
        maxk = Ik.max()
        muk = Bk.mean()

        Ik = Ik - omega[k]*(maxk-muk)/2 # equation (12)

        # Conversion des valeurs de la matrice en pixel (0-255)
        mx,mn = Ik.max(),Ik.min()
        b_sup,b_inf = min(255,mx),max(0,mn)
        Ik = (Ik-mn)*(b_sup-b_inf)/(mx-mn) + b_inf
        
        rgb.append(convolve2d(Ik,H4,'same'))

    return np.dstack((rgb[0],rgb[1],rgb[2])).astype(np.uint8)

File: Codes/Matlab/blind.m
function I_tilde = blind(I)

omega = [0.4,0.35,0.25];

I_tilde = I;

for k=1:3
    maxk = max(max(I(:,:,k)));
    muk = mean(mean(I(:,:,k)));
    I_tilde(:,:,k) = I_tilde(:,:,k) - omega(k)*( maxk - muk)/2;
end

H4 = ones(3,3)/9;
I_tilde = gfilter(I_tilde,H4);

end

